MSI-X support

cagatay-y opened the pull request:
cloud-hypervisor only supports MSI-X interrupts for PCI devices, so support for MSI-X is needed to support running on it. Additionally, MSI-X  allows us to set separate interrupt handlers for the configuration change interrupts and queue updates (even per-queue handlers once we have multiple queues in the future).

~The implementation does not work on QEMU with the TAP network device (tested with `cargo xtask ci rs --arch x86_64 --profile dev  --package httpd --features ci,hermit/virtio-net,hermit/msix qemu --accel --sudo --devices virtio-net-pci --tap`) and possibly other platforms as well. It works with QEMU user network device in addition to cloud-hypervisor, by adding the `hermit/msix` feature to the regular run command.~

mkroening commented on src/drivers/pci.rs:
```diff
@@ -537,6 +537,30 @@ pub(crate) fn init() {
+	addr_high: u32,
+	data: u32,
+	control: u32,
+}
```
This is an entry in the MSI-X table, yes? It might make sense to document that more explicitly:


```suggestion
/// MSI-X Table entry.
#[repr(C)]
pub(crate) struct MsixTableEntry {
    /// Message Address
	message_address: u32,
	
	/// Message Upper Address
	message_upper_address: u32,
	
	/// Message Data
	message_data: u32,
	
	/// Vector Control
	vector_control: u32,
}
```

mkroening commented on src/drivers/net/virtio/pci.rs:
```diff
@@ -44,14 +46,21 @@ impl VirtioNetDriver<Uninit> {
 
+		let irq = device.get_irq();
+		if irq.is_none() {
+			warn!("No interrupt lanes found for virtio-net.");
```
Should this not be an error instead?

mkroening commented on src/drivers/virtio/transport/pci.rs:
```diff
@@ -256,6 +262,19 @@ impl VqCfgHandler<'_> {
+		feature = "virtio-net",
+		not(feature = "rtl8139"),
+		target_arch = "x86_64"
+	))]
```
Why gate this on `not(feature = "rtl8139")`?
And why gate this on `feature = "virtio-net"`? Should we not enable it for all virtio devices?

mkroening reviewed (changes requested):
Thanks! :)

Could you also add how to run this with QEMU user networking and cloud hypervisor to the PR description?

Is this tested in CI already?

cagatay-y committed:
feature(virtio): add message signaled interrupt (MSI-X) support

MSI-X is required to support cloud-hypervisor, as legacy interrupts are
not supported. Additionally, it allows us to set separate handlers for
device configuration interrupts and interrupts of each queue (though
currently we only operate on the queue type granularity). It also seems
to have higher performance.

mkroening commented on src/drivers/pci.rs:
```diff
@@ -463,6 +463,67 @@ pub(crate) fn init(handlers: &mut InterruptHandlerMap) {
+			let msix_entry = unsafe {
+				msix_table
+					.as_mut_ptr()
+					.map(|table| table.cast::<_>().offset(index.try_into().unwrap()))
```
Either specify the type or remove the turbofish:

```suggestion
					.map(|table| table.cast().offset(index.try_into().unwrap()))
```

mkroening commented on src/drivers/fs/mod.rs:
```diff
@@ -159,11 +160,32 @@ impl VirtioFsDriver {
+							driver.lock().handle_device_configuration_interrupt();
+						};
+					},
+					super::pci::msix::NO_QUEUE_HANDLERS.into_iter(),
```
I think we can remove the constant and use a zero-sized iterator instead:

```suggestion
					iter::empty(),
```

cagatay-y replied:
This still requires something like `iter::empty::<(iter::Empty<_>, _)>()` because the compiler cannot deduce a type.

mkroening replied:
I see. Let's do that, then.

mkroening commented on src/drivers/pci.rs:
```diff
@@ -463,6 +463,67 @@ pub(crate) fn init(handlers: &mut InterruptHandlerMap) {
+			msix_table: &mut volatile::VolatileRef<'_, [Self]>,
+			index: u16,
+			vector: u8,
+		) {
```
Could you add a small doc comment?

mkroening replied:
Maybe it would make sense to make this a method taking `self` instead, by implementing an extension trait over `VolatilePtr<'_, [TableEntry], WriteOnly>`?

mkroening reviewed (changes requested):
Thanks, this looks great!

Please ensure that MSI-X is written consistently in code (`MsiX` and `msi_x`).

Also, this seems to be tested successfully with the current CI, right? Could you also add an info log that says whether we are currently using legacy interrupts or MSI-X? :)

mkroening reviewed (approved):
Thank you! 🥳 

The feature and the speedup are great!
